src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,343 @@
+
+    // between genesis and 2100-01-01
+    const int64_t block_start_time = fuzzed_data_provider.ConsumeIntegralInRange<uint32_t>(1231006505, 4102444800);
+
+    // too many blocks at 10min each might cause uint32_t time to overflow if
+    // block_start_time is at the end of the range above
+    assert(n_blocks < 320000);
+
+    // what values for version will we use to signal / not signal?
+    int32_t ver_signal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+    int32_t ver_nosignal = fuzzed_data_provider.ConsumeIntegral<int32_t>();
+
+    // select deployment parameters: bit, start time, timeout
+    int bit = fuzzed_data_provider.ConsumeIntegralInRange<int>(0, VERSIONBITS_NUM_BITS-1);
+
+    bool always_active_test = false;
+    bool never_active_test = false;
+    int64_t start_time;
+    int64_t timeout;
+    if (fuzzed_data_provider.ConsumeBool()) {
+        start_time = block_start_time + start_block*600;
+        timeout = block_start_time + end_block*600;
+
+        assert(start_time <= timeout);
+
+        // allow for times to not exactly match a block
+        if (fuzzed_data_provider.ConsumeBool()) start_time += 300;
+        if (fuzzed_data_provider.ConsumeBool()) timeout += 300;
+
+        // this may make timeout too early; if so, don't run the test
+        if (start_time > timeout) return;
+    } else {
+        if (fuzzed_data_provider.ConsumeBool()) {
+            start_time = Consensus::BIP9Deployment::ALWAYS_ACTIVE;
+            timeout = Consensus::BIP9Deployment::NO_TIMEOUT;
+            always_active_test = true;
+        } else {
+            start_time = 1199145601; // January 1, 2008
+            timeout = 1230767999; // December 31, 2008
+            never_active_test = true;
+        }
+    }
+
+    TestConditionChecker checker(start_time, timeout, period, threshold, bit);
+
+    // Early exit if the versions don't signal sensibly for the deployment
+    if (!checker.Condition(ver_signal)) return;
+    if (checker.Condition(ver_nosignal)) return;
+    if (ver_nosignal < 0) return;
+
+    // TOP_BITS should ensure version will be positive
+    assert(ver_signal > 0);
+
+    // Now that we have chosen time and versions, setup to mine blocks
+    Blocks blocks(block_start_time, ver_signal, ver_nosignal);
+
+    /* Strategy:
+     *  * we will mine a final period worth of blocks, with
+     *    randomised signalling according to a mask
+     *  * but before we mine those blocks, we will mine some
+     *    randomised number of prior periods; with either all
+     *    or no blocks in the period signalling
+     *
+     * We establish the mask first, then consume "bools" until
+     * we run out of fuzz data to work out how many prior periods
+     * there are and which ones will signal.
+     */
+
+    // establish the mask
+    uint32_t signalling_mask = fuzzed_data_provider.ConsumeIntegral<uint32_t>();
+
+    // mine prior periods
+    while (fuzzed_data_provider.remaining_bytes() > 0) {
```

Comment:
Could replace with 
```suggestion
    while (fuzzed_data_provider.ConsumeBool()) {
```

or mention that the provider is used up after this call and should not be accessed again?

Reply:
Couldn't the subsequent `ConsumeBool()` fail if this is the last byte in the data provider?

Reply:
Added the comment

Reply:
FuzzedDataProvider itself doesn't fail if the last byte was consumed. It will simply return constant values. In the case of bool, `false`. So using `ConsumeBool()` would be identical to `remaining_bytes`, with the difference that the fuzz input is one more byte large (to encode the consumed bool) per iteration. Also, `ConsumeBool` would leave open the possibility to use the data provider afterwards without invalidating the inputs.

Reply:
Ah, thanks Marco. I agree that this would be better as ConsumeBool (and then we could move the `signalling_mask` declaration down to where it's used).

Reply:
Tried the `while (ConsumeBool()) { signal = ConsumeBool(); } mask = ConsumeIntegral();` alternative and it seems like the current approach gets more coverage more quickly (1896 cov 6382 ft after 20s and 1898 cov 7041 ft after ~3m; vs 1893 cov 6111 ft after 20s and 1895 cov 6960 ft after ~3m), so am leaving this as-is.

Reply:
I am trying to reproduce with `FUZZ=versionbits ./src/test/fuzz/fuzz -use_value_profile=1 -entropic=1 -max_total_time=10`.
Looks like I am exhausting line coverage after 10 seconds with `ConsumeBool()`:

```
#68347	DONE   cov: 653 ft: 3266 corp: 557/16057b lim: 122 exec/s: 6213 rss: 46Mb
```


Whereas remaining_bytes gives lower line coverage after 10 seconds:

```
#58568	DONE   cov: 650 ft: 3262 corp: 530/14599b lim: 128 exec/s: 5324 rss: 46Mb
```

In either case, this seems premature optimization. A time-scale of 10 seconds is nothing compared to the CPU time that is going to be spent on this in our fuzz farms. I think the primary thing to optimize is to write tests in a way they are easy to understand and hard to get wrong when modified/written. For example, in another test (commit fa42da2d5424c0aeccfae4b49fde2bea330b63dc) a fully consumed buffer has already bitten me.

Reply:
Well, you're getting about 150x the exec/s I am, so getting thorough coverage in 10s when it takes me 2m makes sense. I don't follow why your "cov" figures are 650 rather than the 1900 or so I get though?

Reply:
"cov" and "ft" depend on the instrumentation (version of libFuzzer, sanitizers, run time flags, ...), they are not considered a "stable interface".

Even if it took 2 minutes to saturate line coverage, I'd still consider it more than acceptable. We have many targets that are ground with hundreds of CPU hours and still aren't close to being saturated in line coverage.